lib/devise-security/models/password_expirable.rb

```diff
@@ -4,66 +4,83 @@
 
 module Devise
   module Models
-
-    # PasswordExpirable takes care of change password after
+    # PasswordExpirable makes passwords expire after a configurable amount of
+    # time, or on demand
     module PasswordExpirable
       extend ActiveSupport::Concern
 
       included do
         before_save :update_password_changed
       end
 
-      # is an password change required?
+      # Is a password change required?
+      # @return [Boolean]
+      # @return [true] if +password_changed_at+ has not been set or if it is old
+      #   enough based on +expire_password_after+ configuration.
       def need_change_password?
-        if expired_password_after_numeric?
-          self.password_changed_at.nil? || self.password_changed_at < self.expire_password_after.seconds.ago
+        if enabled?
+          if expire_on_demand?
+            # Only expire when password_changed_at is nil
+            password_changed_at.nil?
+          else
+            # Expire when password_changed_at is nil or when the last time the
+            # password was changed is sufficiently old
+            password_changed_at.nil? || password_changed_at < expire_password_after.seconds.ago
+          end
         else
           false
         end
       end
 
-      # set a fake datetime so a password change is needed and save the record
+      # Adjust the +password_changed_at+ field so that it will require a
+      # password update and save the record (without validations)
+      # @return [Boolean]
       def need_change_password!
-        if expired_password_after_numeric?
-          need_change_password
-          self.save(validate: false)
-        end
+        return unless enabled?
+        need_change_password
+        save(validate: false)
       end
 
-      # set a fake datetime so a password change is needed
+      # Clear the +password_changed_at+ field so that it will require a
+      # password update.
+      # @return [void]
       def need_change_password
-        if expired_password_after_numeric?
-          self.password_changed_at = self.expire_password_after.seconds.ago
-        end
-
-        # is date not set it will set default to need set new password next login
-        need_change_password if self.password_changed_at.nil?
-
-        self.password_changed_at
+        return unless enabled?
+        self.password_changed_at = nil
```

Comment:
Seems like this is a logic change from the original, not just cleaning up CC. Is that intentional?

Reply:
This is a slightly different implementation.  The original would make up a fake time that would cause the record to appear as if it needed to be changed.  I didn't particularly care for that because it makes it look like the user changed their password at a time when they really didn't and it makes it impossible to distinguish between an administrative expiration of the password and a real one.  Just setting the `password_changed_at` to `nil` has the same effect and it's possible to tell that the password was deliberately expired because that is the only way to get a `nil` in that column.